hazelcast/serialization/predicate.py

```diff
@@ -204,6 +205,133 @@ def write_data(self, object_data_output):
     def __repr__(self):
         return "TruePredicate()"
 
+
+class PagingPredicate(Predicate):
+    CLASS_ID = 15
+
+    __NULL_ANCHOR = (-1, None)
+
+    def __init__(self, predicate, page_size, comparator=None):
+        """
+        Creates a Paging predicate with provided page size, internal predicate, and optional comparator.
+        :param predicate: predicate to filter the results
+        :param page_size: page size of each result set
+        :param comparator: (Optional) a comparator object used to sort the results.
+            Defines compare method on (K,V) tuples. Must be an implementation of hazelcast.core.Comparator
+            WARNING: comparator must extend Comparator, and either IdentifiedDataSerializable or Portable.
+        """
+        if isinstance(predicate, PagingPredicate):
+            raise TypeError('Nested paging predicate not supported.')
+        self.internal_predicate = predicate
+        self.comparator = comparator
+        if page_size <= 0:
+            raise ValueError('page_size should be greater than 0.')
+        self.page_size = page_size
+        self.page = 0  # initialized to be on first page
+        self.iteration_type = ITERATION_TYPE.ENTRY  # ENTRY as default.
+        self.anchor_list = []  # List of pairs: (nearest page, (anchor key, anchor value))
+
+    def __repr__(self):
+        return "PagingPredicate(predicate=%s, page_size=%s, comparator=%s)" % (self.internal_predicate,
+                                                                               self.page_size, self.comparator)
+
+    def write_data(self, output):
+        output.write_object(self.internal_predicate)
+        output.write_object(self.comparator)
+        output.write_int(self.page)
+        output.write_int(self.page_size)
+        output.write_utf(ITERATION_TYPE.reverse.get(self.iteration_type, None))
+        output.write_int(len(self.anchor_list))
+        for nearest_page, (anchor_key, anchor_value) in self.anchor_list:
+            output.write_int(nearest_page)
+            output.write_object(anchor_key)
+            output.write_object(anchor_value)
+
+    def next_page(self):
+        """
+        Sets page index to next page. If new index is out of range, the query results that this paging predicate will
+        retrieve will be an empty list.
+        :return (int) current page index
+        """
+        self.page += 1
+        return self.page
+
+    def previous_page(self):
+        """
+        If current page index is 0, this method does nothing. Otherwise, it sets page index to previous page.
+        :return (int) current page index
+        """
+        if self.page != 0:
+            self.page -= 1
+        return self.page
+
+    def set_page(self, page_no):
```

Comment:
Minor: these getters and setters are not very 'pythonic'. If they are simply accessed without further checks, (like `iteration_type`) I believe making it public and getting rid of `get_x` and `set_x` should be enough.

If there are some checks (like `set_page`), we could define the field as protected and define properties for getters and setters (something like https://stackoverflow.com/a/2627034/10236486)

What do you think about it?

Reply:
Oh, that's a good idea. I mainly tried to keep the getters and setters to be consistent with the reference manual, docs, and the Hazelcast API on the Java side. I agree we can get rid of the getters and setters for stuff like `comparator` and `iteration_type`. But I think we should keep the getter for `page_size` and make the field protected since we don't want to allow the user the change `page_size` after initialization. And I would keep `set_page` because I think it is a more intuitive way to request page 100, for example, and more consistent across clients. But if you think we should go with the more pythonic way, I am happy to implement it. What do you think?

Reply:
That looks pretty good to me. Thanks for the changes